### PR TITLE
58-fix-video-recorder

### LINKED ISSUE #58: Fix video recorder
### Describe the bug
Now that we changed the sampling of trajectories, we need to correct where the replay starts.

Screenshot 2024-12-21 230631

Theses trajectories shouldn't all start at the reset postition.

### PR DESCRIPTION
**Pull request by Wizzardexe**
**Issue:** Fix #58

**What?**
- Video generation starts in the wrong state

**How?**
- Save all states properly

### CODE REVIEW

File: framework/reward_net.py [hunk 1]
```diff
@@ -116,14 +116,14 @@ def preference_hat_entropy_member(self, traj1, traj2, member: int = -1):
             traj2.samples.actions,
             member=member,
         )
-        r_hat1 = r1.sum(axis=1)
-        r_hat2 = r2.sum(axis=1)
-        r_hat = torch.cat([r_hat1, r_hat2], axis=-1)
```
Comment (Wizzardexe) [inline framework/reward_net.py:121]: Fixed the type issue in torch.cat()

File: framework/replay_buffer.py [hunk 2]
```diff
@@ -96,12 +104,12 @@ def sample_trajectories(
         :return: two lists of mb_size many trajectories
         """
 
-        indices = np.random.choice(self.pos, 2 * mb_size, replace=False)
+        indices = np.random.choice(self.pos - 32, 2 * mb_size, replace=False)
 
         trajectories = [
             self.get_trajectory(
                 start,
-                (start + traj_len) % self.pos,
```
Comment (Wizzardexe) [inline framework/replay_buffer.py:104]: This allowed us to generate trajectories where start > end

**Comment by Wizzardexe** [comment]
Before: 
Screenshot 2024-12-22 014554
Notice that the before pictures are 30k later in the training.

After:
Screenshot 2024-12-22 014522

My implementation only works with a single env rightnow, but I will implement multiple soon.

File: framework/replay_buffer.py (hunk 2 above)
Comment (Wizzardexe) [reply framework/replay_buffer.py:104]: `trajectory_indices = np.arange(start_idx, end_idx)` Output: [] 
`get_trajectory()` returns empty tensors

File: framework/replay_buffer.py [hunk 3]
```diff
@@ -115,14 +115,13 @@ def sample_trajectories(
...
         logging.debug(f"trajectory length: {traj_len}, ")
-
         return trajectories[:mb_size], trajectories[mb_size:]
 
     def _get_samples(
         self, batch_inds: np.ndarray, env: Optional[VecNormalize] = None
     ) -> ReplayBufferSampleHF:
         # Sample randomly the env idx
-        env_indices = np.random.randint(0, high=self.n_envs, size=(len(batch_inds),))
+        env_indices = np.random.randint(0, high=self.n_envs)
```
Comment (Wizzardexe) [inline framework/replay_buffer.py:125]: I changed this to only one env, because we else generate trajectories which are not comparable for humans.

**Review by bkuen — APPROVED** [review]

**Pull request merged**